File: automation/extension/src/worker/authorization.ts
import type { RawOrder, ScrapeControl } from '../shared/contract'

// 扩展侧统一授权守卫（P0 2026-06-22）。
//
// 单一判据来源：服务端 GET /inbox/scrape/control 返回的 ScrapeControl.authorized（由 inbox-service
// 据任务租约 + 服务端时钟算出）。扩展所有副作用入口（自动抓取循环、翻页、推送、打标）都过这里，
// 任何一处都不另写判断、也绝不信任 chrome.storage / localStorage / 内存里残留的 enabled/running/旧任务。
//
// fail-closed：control 为空 / authorized 非 true / 缺时间范围 → 一律未授权；订单缺付款时间 / 越窗 → 不在范围。

export const MARK_ACTION = 'mark'
export const SCRAPE_ACTION = 'scrape'

/** 当前是否被授权执行（唯一判据）。旧服务无 authorized 字段 → undefined → false。 */
export function isAuthorized(control: ScrapeControl | null | undefined): boolean {
  return Boolean(control && control.authorized)
}

/** 某具体操作（scrape/mark）当前是否被授权：先要整体授权，再看它在 allowed_actions 里。 */
export function actionAllowed(control: ScrapeControl | null | undefined, action: string): boolean {
  if (!isAuthorized(control)) return false
  const csv = control?.allowed_actions ?? ''
  return csv
    .split(',')
    .map((a) => a.trim())
    .includes(action)
}

/**
 * 把墙钟时间串归一成可比较的数字 YYYYMMDDHHMMSS。
 * 同时吃店小秘付款时间「2026-06-19 02:25」与服务端 ISO「2026-06-19T00:00:00」——
 * 两者都是店小秘墙钟、同域比较即正确；用数字比避免「空格 0x20 < T 0x54」的字典序陷阱。
 */
export function toComparable(value: string | null | undefined): number | null {
  if (!value) return null
  const m = value.match(/(\d{4})\D(\d{1,2})\D(\d{1,2})\D(\d{1,2})\D(\d{2})(?:\D(\d{2}))?/)
  if (!m) return null
  const pad = (s: string) => s.padStart(2, '0')
  return Number(`${m[1]}${pad(m[2])}${pad(m[3])}${pad(m[4])}${pad(m[5])}${pad(m[6] ?? '00')}`)
}

/**
 * 订单是否落在当前任务的**付款时间范围** [scrape_from, scrape_to] 内。
 * fail-closed：未授权 / 无付款时间（无法按时间判定，防历史/未付款单混入）/ 越窗 → false。
 */
export function orderInScope(order: RawOrder, control: ScrapeControl | null | undefined): boolean {
  if (!isAuthorized(control)) return false
  const paid = toComparable(order.paid_at)
  if (paid === null) return false
  const from = toComparable(control?.scrape_from)
  if (from !== null && paid < from) return false
  const to = toComparable(control?.scrape_to)
  if (to !== null && paid > to) return false
  return true
}

/**
 * 本页是否已**整页越过时间窗下界**（所有带付款时间的单都早于 scrape_from）。
 * 店小秘列表按付款时间倒序 → 一旦整页都旧于下界，再往后翻只会是更早的历史单 → 应停止翻页（防回溯扫历史）。
 * 无下界 / 本页没有可判定付款时间的单 → 返回 false（不据此停，交给游标/末页/上限）。
 */
export function pageBelowWindowFloor(
  orders: RawOrder[],
  control: ScrapeControl | null | undefined,
): boolean {
  const from = toComparable(control?.scrape_from)
  if (from === null) return false
  const paids = orders
    .map((o) => toComparable(o.paid_at))
    .filter((n): n is number => n !== null)
  if (paids.length === 0) return false
  return paids.every((p) => p < from)
}

File: automation/inbox-service/app/authorization.py
from __future__ import annotations

# 任务授权（统一守卫，P0 2026-06-22）。
#
# 背景：原来扩展是否执行只看 ScrapeControl.enabled 一个布尔，而它存在独立常驻的 inbox-service DB——
# flower 关掉后仍永久 true，扩展每次开店小秘页就误以为已授权，自动抓取+打标历史订单（副作用失控）。
#
# 修复：enabled 不再等于授权。授权 = 存在**未过期的任务租约**（flower 持续心跳续约）。这里是
# 「当前是否有有效授权 / 某订单是否在授权范围内」的唯一判定处——服务端各副作用端点都调它，扩展也读
# GET /scrape/control 返回的 authorized（由本模块算）。fail-closed：任何字段缺失/过期/无法判定 → 拒绝。

from datetime import datetime, timezone

from app.models import ScrapeControl

ACTION_SCRAPE = "scrape"
ACTION_MARK = "mark"


def now_naive() -> datetime:
    """当前 UTC（naive）。租约/付款时间都按 naive UTC 存比，避免与 SQLite 读回的 naive 值混 tz。"""
    return datetime.now(timezone.utc).replace(tzinfo=None)


def _as_naive(value: datetime | None) -> datetime | None:
    if value is None:
        return None
    if value.tzinfo is not None:
        return value.astimezone(timezone.utc).replace(tzinfo=None)
    return value


def lease_valid(control: ScrapeControl | None, now: datetime | None = None) -> bool:
    """是否存在**未过期**的任务租约（有 task_id 且 lease_expires_at 在未来）。

    这是服务端「现在是否有 flower 在管」的核心判据——副作用端点据它 fail-closed。
    不看 enabled / scrape_from（那些归 is_authorized）；租约本身在不在、过没过期是第一道闸。
    """
    if control is None:
        return False
    if not control.task_id:
        return False
    expires = _as_naive(control.lease_expires_at)
    if expires is None:
        return False
    return (now or now_naive()) < expires


def is_authorized(control: ScrapeControl | None, now: datetime | None = None) -> bool:
    """扩展是否被授权执行自动抓取/打标。

    = enabled 且 租约有效 且 scrape_from 有值（订单时间范围必填——规范：缺时间范围必须拒绝）。
    """
    if control is None or not control.enabled:
        return False
    if not lease_valid(control, now):
        return False
    if control.scrape_from is None:  # 没有订单时间范围 → 拒绝（绝不默认全量）
        return False
    return True


def allowed_action_set(control: ScrapeControl | None) -> set[str]:
    """任务允许的操作集合。无任务/未配置 → 空集（什么都不允许，fail-closed）。"""
    if control is None or not control.allowed_actions:
        return set()
    return {a.strip() for a in control.allowed_actions.split(",") if a.strip()}


def action_allowed(control: ScrapeControl | None, action: str, now: datetime | None = None) -> bool:
    """某具体操作（scrape/mark）当前是否被授权：先要整体授权，再看该操作在 allowed_actions 里。"""
    if not is_authorized(control, now):
        return False
    return action in allowed_action_set(control)


def shop_scope_set(control: ScrapeControl | None) -> set[str] | None:
    """允许的店铺集合；None=不限店铺。空白 → None。"""
    if control is None or not control.shop_scope:
        return None
    shops = {s.strip() for s in control.shop_scope.split(",") if s.strip()}
    return shops or None


def paid_in_time_window(control: ScrapeControl | None, paid_at: datetime | None) -> bool:
    """仅按**付款时间窗** [scrape_from, scrape_to] 判定（不看店铺）。

    供 /scrape/diff 规划阶段用（清单只有 order_id+付款时间、没有店铺）；店铺过滤留到入库 order_in_scope。
    fail-closed：未授权 / paid_at 缺失 / 不在窗内 → False。
    """
    if not is_authorized(control):
        return False
    assert control is not None
    paid = _as_naive(paid_at)
    if paid is None:
        return False
    start = _as_naive(control.scrape_from)
    if start is not None and paid < start:
        return False
    end = _as_naive(control.scrape_to)
    if end is not None and paid > end:
        return False
    return True


def order_in_scope(
    control: ScrapeControl | None,
    *,
    paid_at: datetime | None,
    shop: str | None = None,
) -> bool:
    """订单是否落在当前任务的范围内（付款时间窗 [scrape_from, scrape_to] + 店铺）。

    fail-closed：
    - 未授权 → False。
    - paid_at 缺失（无法按时间判定）→ False（绝不把无付款时间的单当「在范围内」，防历史/未付款单混入）。
    - paid_at < scrape_from 或 > scrape_to → False。
    - shop_scope 非空且订单店铺不在其中 → False。
    """
    if not is_authorized(control):
        return False
    assert control is not None  # is_authorized 已保证
    paid = _as_naive(paid_at)
    if paid is None:
        return False
    start = _as_naive(control.scrape_from)
    if start is not None and paid < start:
        return False
    end = _as_naive(control.scrape_to)
    if end is not None and paid > end:
        return False
    shops = shop_scope_set(control)
    if shops is not None and (shop or "") not in shops:
        return False
    return True
